### ISSUE TITLE
Merge into ReOpenLDAP

### ISSUE THREAD

**Issue by erthink** [osstech-jp/openldap#6]
I offer to join the [ReOpenLDAP](https://github.com/leo-yuriev/ReOpenLDAP) and merge all of your improvements.

In ReOpenLDAP fixed a lot of problems, especially in the replication code.
Here a [few slides](https://ldapcon.org/2017/wp-content/uploads/2017/08/1_ReOpenLDAP-LDAPCON-2017.pdf) about ReOpenLDAP from [LDAPCon-2017](https://ldapcon.org/2017/conference-program/)

If necessary, support AIX can be returned (I deleted it because it was not necessary for us and the opportunity to test).

Be free to ask a question or make a pull-request.

**Comment by hamano** [comment]
I'm trying to build back-wt with ReOpenLDAP, but it seems to be difficult to maintain common code with OpenLDAP.
The back-wt is in active development stage, so I will merge it when it become stabilized.

**Comment by erthink** [comment]
@hamano, please point me to some semistable release/branch - I wanna try to merge it.

**Comment by hamano** [comment]
This is stabe version of back-wt:
https://github.com/osstech-jp/openldap/tree/wiredtiger

And this is just a configure update, and I tried integrating back-wt directory (servers/slapd/back-wt) by git submodule.
https://github.com/osstech-jp/ReOpenLDAP/tree/wiredtiger

However, there are too many code to be changes. . .

**Comment by hamano** [comment]
this is build instruction I tried:

~~~
wget http://source.wiredtiger.com/releases/wiredtiger-3.0.0.tar.bz2
tar xf wiredtiger-3.0.0.tar.bz2
cd wiredtiger-3.0.0
configure
make
make install
~~~

If you prefer static linking with wiredtiger.
edit /usr/local/lib/pkgconfig/wiredtiger.pc:
~~~
< Libs: -L${libdir} -lwiredtiger
---
> Libs: -L${libdir} -lrt -ldl ${libdir}/libwiredtiger.a
~~~

~~~
git clone ReOpenLDAP
cd ReOpenLDAP
git submodule add -b wiredtiger git@github.com:osstech-jp/openldap.git contrib/wiredtiger
ln -sf ../../contrib/wiredtiger/servers/slapd/back-wt/ ./servers/slapd/back-wt
./bootstrap.sh --dont-cleanup
./configure --enable-wt
~~~

**Comment by erthink** [comment]
@hamano, please verify changes, especially [this line](https://github.com/leo-yuriev/ReOpenLDAP/commit/f38625ec2ed917f2c0aacbd1d2fa87c9c7b0961d#diff-a5381309b37d6a0816ad21022f8ab6e4R156).

Regards.

**Comment by hamano** [comment]
@leo-yuriev

~~~
 +	if (rc == WT_NOTFOUND ||
￼ +			/* FIXME: @hamano - verify braces */
￼ +		(!manageDSAit && e && is_entry_glue(e)) ) {
~~~

This changes comes from here: https://github.com/osstech-jp/openldap/commit/62f0d3e30327152aea39025a0dac35f3caa4ebf8

Unfortunately I forgot the details, but this fix is for test032-chain and test033-glue-syncrepl.
Our testing CI system records that these tests are successful after applied the changes.
Currently, the test result of the back-wt is 57 / 63.
Have a try test:

~~~
./run -b wt all
~~~

Thanks!

**Comment by hamano** [comment]
@leo-yuriev Could you specify compile flag -DREOPENLDAP?
I'd like to keep code as same as possible between OpenLDAP and ReOpenLDAP.

~~~
# ifdef REOPENLDAP
#include "reldap.h"
#else
#include "portable.h"
#endif
~~~

I have no idea about Debug() vargs...

**Comment by erthink** [comment]
> @leo-yuriev Could you specify compile flag -DREOPENLDAP?
I'd like to keep code as same as possible between OpenLDAP and ReOpenLDAP.

I don't think you'd ask me if you saw how many bugs were fixed in slapd.
Technically this could be done by `diff --ifdef=REOPENLDAP`, but firstly bugs should be fixed...

> I have no idea about Debug() vargs...

This is a trouble.

**Comment by erthink** [comment]
```
	if (rc == WT_NOTFOUND ||
		(!manageDSAit && e && is_entry_glue(e)) ) {
```
@hamano, your code has a bug. Such braces exist in the `compare.c`, but missing in the `delete.c`.
Moreover, there are bugs due absent function prototypes (mismatched arg types, i.e. `wt_ctx` and `wt_ctx->session`).

I fixed this while this merge.

**Comment by erthink** [comment]
One more - [`wt_hasSubordinates()`](https://github.com/leo-yuriev/ReOpenLDAP/commit/a61a4146cec14bb17593ae90f020c58082884cd6#diff-c8f04a248ed57467dccd97db67af966eR339) was missed.

**Comment by hamano** [comment]
@leo-yuriev I misunderstood your words: parenthesis and brace.
I just fixed cb8b7d4062e3ac0b0ae4b34a81503b7d1bb3ae49 and I backported important bugs 47bd0c5638e903723ca891cba221a9713e4e8fa1

thanks.
